#Problem Statement

This dataset is created for the prediction of House Prices from an Indian perspective Towns. The dataset contains various features that are important during the application for the selling the Houses. The predicted output obtained from the classification algorithm gives a fair idea about the chances of a Houses Sold.


<table align="center" width=100%>
    <tr>
        </td>
            </td>
        </td>
        <td>
            <div align="center">
                <font color="#216180" size=8px>
                    <b> Regression modeling on House Price Data
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

In [1]:
# import Pandas
import pandas as pd

#import numpy
import numpy as np

#import subpackage of Matpiotlib
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap

#import Seaborn
import seaborn as sns

# # to supress warnings
# from warnings import filtervarnings. filtersarnings('ignore')



### Load the data state

In [2]:
# Load the data and name it 'df_House_price'

In [3]:
# Load the csv file
# store the data in 'df_houseprice'
df_houseprice =pd.read_csv(r"C:\Users\HP\Downloads\House_Price1.csv")

 #display first file observation                         
df_houseprice.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200       NaN        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

In [4]:
df=pd.read_csv(r"C:\Users\HP\Downloads\House_Price1.csv")
df_houseprice.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200       NaN        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

## Data Overview

In [5]:
# Name the columns 
df_houseprice.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

**Interpretation**

 - `Price`:Price of the house.
 -  `resid_area`:Residential area where my house is located.
   

In [6]:
# Basic Info of the Dataset
df_houseprice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   price        506 non-null    float64
 1   resid_area   506 non-null    float64
 2   air_qual     506 non-null    float64
 3   room_num     506 non-null    float64
 4   age          506 non-null    float64
 5   distance     506 non-null    float64
 6   teachers     506 non-null    float64
 7   poor_prop    506 non-null    float64
 8   airport      506 non-null    object 
 9   n_hos_beds   498 non-null    float64
 10  n_hot_rooms  506 non-null    float64
 11  waterbody    351 non-null    object 
 12  rainfall     506 non-null    int64  
 13  bus_ter      506 non-null    object 
 14  parks        506 non-null    float64
 15  Sold         506 non-null    int64  
dtypes: float64(11), int64(2), object(3)
memory usage: 63.4+ KB


**Interpretation**

- We have 13 numerical columns and 3 categorical columns.
- We have null values in the data.
- We have total 16 columns.

In [7]:
# Shape of the data
df_houseprice.shape

(506, 16)

**Interpretation**
- we have total 16 columns
- we have 506 observations

In [8]:
df_houseprice.describe().T

count       mean        std        min        25%        50%  \
price        506.0  22.528854   9.182176   5.000000  17.025000  21.200000   
resid_area   506.0  41.136779   6.860353  30.460000  35.190000  39.690000   
air_qual     506.0   0.554695   0.115878   0.385000   0.449000   0.538000   
room_num     506.0   6.284634   0.702617   3.561000   5.885500   6.208500   
age          506.0  68.574901  28.148861   2.900000  45.025000  77.500000   
distance     506.0   3.971996   2.108532   1.130000   2.270000   3.385000   
teachers     506.0  21.544466   2.164946  18.000000  19.800000  20.950000   
poor_prop    506.0  12.653063   7.141062   1.730000   6.950000  11.360000   
n_hos_beds   498.0   7.899767   1.476683   5.268000   6.634500   7.999000   
n_hot_rooms  506.0  13.041605   5.238957  10.057600  11.189800  12.720000   
rainfall     506.0  39.181818  12.513697   3.000000  28.000000  39.000000   
parks        506.0   0.054454   0.010632   0.033292   0.046464   0.053507   
Sold         506.0   0.454545   0.498422   0.000000   0.000000   0.000000   

                   75%         max  
price        25.000000   50.000000  
resid_area   48.100000   57.740000  
air_qual      0.624000    0.871000  
room_num      6.623500    8.780000  
age          94.075000  100.000000  
distance      5.367500   12.320000  
teachers     22.600000   27.400000  
poor_prop    16.955000   37.970000  
n_hos_beds    9.088000   10.876000  
n_hot_rooms  14.170800  101.120000  
rainfall     50.000000   60.000000  
parks         0.061397    0.086711  
Sold          1.000000    1.000000

**Interpretation**
- We have outliers in the data, most of the outliers are in age column.

## Data prerocessing.


### a.datacleaning

In [9]:
# show the first five observations.
df_houseprice.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200       NaN        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

In [10]:
# Rename the columns
# Step 1:Show the column names.
df_houseprice.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

In [11]:
# Step 2:Allocate new column names to old column names.
# Make the dictionary {'Old column name':'New column name'}
d1={'price':'Price',
  'resid_area':'Resid_Area',
  'air_qual':'Air_Qual',
   'room_num':'Room_Nums',
    'age':'Age',
    'distance':'Distance',
    'teachers':'Teachers',
    'poor_prop':'Preportion_of_poor_population',
    'airport':'Airport',
    'n_hos_beds':'Number_of_hospital_beds',
    'n_hot_rooms':'Number_of_hotel_rooms',
     'waterbody': 'Waterbody',
    'rainfall':'Rainfall',
    'bus_ter':'Bus_Terminal',
    'parks':'Parks',
    'Sold':'sold'
 }

In [12]:
df_houseprice=df_houseprice.rename(columns=d1)


In [13]:
# Authenticate the data
df_houseprice.columns

Index(['Price', 'Resid_Area', 'Air_Qual', 'Room_Nums', 'Age', 'Distance',
       'Teachers', 'Preportion_of_poor_population', 'Airport',
       'Number_of_hospital_beds', 'Number_of_hotel_rooms', 'Waterbody',
       'Rainfall', 'Bus_Terminal', 'Parks', 'sold'],
      dtype='object')

### b.Null value handling

In [14]:
# Find the valid null values
df_houseprice.isnull().sum()

Price                              0
Resid_Area                         0
Air_Qual                           0
Room_Nums                          0
Age                                0
Distance                           0
Teachers                           0
Preportion_of_poor_population      0
Airport                            0
Number_of_hospital_beds            8
Number_of_hotel_rooms              0
Waterbody                        155
Rainfall                           0
Bus_Terminal                       0
Parks                              0
sold                               0
dtype: int64

In [15]:
# percentage missing value
round(df_houseprice.isnull().sum()/len(df_houseprice)*100,3)

Price                             0.000
Resid_Area                        0.000
Air_Qual                          0.000
Room_Nums                         0.000
Age                               0.000
Distance                          0.000
Teachers                          0.000
Preportion_of_poor_population     0.000
Airport                           0.000
Number_of_hospital_beds           1.581
Number_of_hotel_rooms             0.000
Waterbody                        30.632
Rainfall                          0.000
Bus_Terminal                      0.000
Parks                             0.000
sold                              0.000
dtype: float64

In [16]:
# Find the nature of the column
df_houseprice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Price                          506 non-null    float64
 1   Resid_Area                     506 non-null    float64
 2   Air_Qual                       506 non-null    float64
 3   Room_Nums                      506 non-null    float64
 4   Age                            506 non-null    float64
 5   Distance                       506 non-null    float64
 6   Teachers                       506 non-null    float64
 7   Preportion_of_poor_population  506 non-null    float64
 8   Airport                        506 non-null    object 
 9   Number_of_hospital_beds        498 non-null    float64
 10  Number_of_hotel_rooms          506 non-null    float64
 11  Waterbody                      351 non-null    object 
 12  Rainfall                       506 non-null    int

In [17]:
# This is a categorical column so we can replace the null with mode.
df_houseprice.Waterbody.mode()

0    River
Name: Waterbody, dtype: object

In [18]:
# Replace the null in that column
df_houseprice.Waterbody=df_houseprice.Waterbody.fillna('River')

In [19]:
# In Number_of_hospital_beds contains less than 10% missing records
df_houseprice.dropna(inplace=True)

In [20]:
round(df_houseprice.isnull().sum()/len(df_houseprice)*100,3)

Price                            0.0
Resid_Area                       0.0
Air_Qual                         0.0
Room_Nums                        0.0
Age                              0.0
Distance                         0.0
Teachers                         0.0
Preportion_of_poor_population    0.0
Airport                          0.0
Number_of_hospital_beds          0.0
Number_of_hotel_rooms            0.0
Waterbody                        0.0
Rainfall                         0.0
Bus_Terminal                     0.0
Parks                            0.0
sold                             0.0
dtype: float64

In [21]:
target = df_houseprice['sold']

In [22]:
df_Independant = df_houseprice.drop('sold', axis =1)

In [23]:
df_cat =df_Independant.select_dtypes(include='object')

In [24]:
df_cat.head()

Airport Waterbody Bus_Terminal
0     YES     River          YES
1      NO      Lake          YES
2      NO     River          YES
3     YES      Lake          YES
4      NO      Lake          YES

In [25]:
encoded =pd.get_dummies(df_cat)
encoded = encoded.reset_index()

In [26]:
encoded.head()

index  Airport_NO  Airport_YES  Waterbody_Lake  Waterbody_Lake and River  \
0      0       False         True           False                     False   
1      1        True        False            True                     False   
2      2        True        False           False                     False   
3      3       False         True            True                     False   
4      4        True        False            True                     False   

   Waterbody_River  Bus_Terminal_YES  
0             True              True  
1            False              True  
2             True              True  
3            False              True  
4            False              True

## Scale the numerical Data 

In [27]:
df_num = df_Independant.select_dtypes(include = 'number')

In [28]:
from sklearn.preprocessing import StandardScaler

In [29]:
ss = StandardScaler()

In [30]:
scale = ss.fit_transform(df_num)

In [31]:
print(scale)

[[ 0.15479189 -1.28244435 -0.13709359 ... -0.35211111 -1.30206786
  -0.47313205]
 [-0.10659165 -0.59005882 -0.73319492 ... -0.16601943  0.21519397
  -0.77522463]
 [ 1.32012686 -0.59005882 -0.73319492 ... 16.71034024 -0.10422957
  -0.81123666]
 ...
 [ 0.14390091  0.11687263  0.16527665 ... -0.16252832 -0.66322077
   0.30293265]
 [-0.06302773  0.11687263  0.16527665 ...  0.40379146  0.6144734
   0.59750894]
 [-0.38975716  0.11687263  0.16527665 ... -0.54943507  0.45476163
   0.56369225]]


In [32]:
df_scale = pd.DataFrame(scale,columns = df_num.columns)

In [33]:
df_scale = df_scale.reset_index()
df_scale

index     Price  Resid_Area  Air_Qual  Room_Nums       Age  Distance  \
0        0  0.154792   -1.282444 -0.137094   0.408975 -0.115644  0.172845   
1        1 -0.106592   -0.590059 -0.733195   0.189974  0.369741  0.476100   
2        2  1.320127   -0.590059 -0.733195   1.276446 -0.260905  0.495053   
3        3  1.178544   -1.301354 -0.828226   1.010516 -0.802977  1.054179   
4        4  1.483492   -1.301354 -0.828226   1.222407 -0.505369  1.030487   
..     ...       ...         ...       ...        ...       ...       ...   
493    493 -0.019464    0.116873  0.165277   0.434572  0.022531 -0.637413   
494    494 -0.215501    0.116873  0.165277  -0.238073  0.291796 -0.732180   
495    495  0.143901    0.116873  0.165277   0.979230  0.798438 -0.779564   
496    496 -0.063028    0.116873  0.165277   0.720411  0.738208 -0.684797   
497    497 -0.389757    0.116873  0.165277  -0.366061  0.437057 -0.599506   

     Teachers  Preportion_of_poor_population  Number_of_hospital_beds  \
0    1.462167                      -1.066368                -1.640299   
1    0.304963                      -0.484136                -0.384875   
2    0.304963                      -1.199330                -0.342847   
3   -0.111631                      -1.351886                 0.927490   
4   -0.111631                      -1.017382                 0.626514   
..        ...                            ...                      ...   
493 -1.176258                      -0.409958                 0.981720   
494 -1.176258                      -0.492534                -0.872945   
495 -1.176258                      -0.973995                -1.641654   
496 -1.176258                      -0.856429                 0.027273   
497 -1.176258                      -0.660486                 1.613499   

     Number_of_hotel_rooms  Rainfall     Parks  
0                -0.352111 -1.302068 -0.473132  
1                -0.166019  0.215194 -0.775225  
2                16.710340 -0.104230 -0.811237  
3                -0.337843  0.454762 -0.680402  
4                -0.333593  1.253320 -1.404723  
..                     ...       ...       ...  
493              -0.164805 -0.982644  0.155116  
494               0.022197 -1.541636  0.522886  
495              -0.162528 -0.663221  0.302933  
496               0.403791  0.614473  0.597509  
497              -0.549435  0.454762  0.563692  

[498 rows x 13 columns]

In [34]:
X = pd.concat([df_scale,encoded] , axis = 1)

In [35]:
X.head()

index     Price  Resid_Area  Air_Qual  Room_Nums       Age  Distance  \
0      0  0.154792   -1.282444 -0.137094   0.408975 -0.115644  0.172845   
1      1 -0.106592   -0.590059 -0.733195   0.189974  0.369741  0.476100   
2      2  1.320127   -0.590059 -0.733195   1.276446 -0.260905  0.495053   
3      3  1.178544   -1.301354 -0.828226   1.010516 -0.802977  1.054179   
4      4  1.483492   -1.301354 -0.828226   1.222407 -0.505369  1.030487   

   Teachers  Preportion_of_poor_population  Number_of_hospital_beds  \
0  1.462167                      -1.066368                -1.640299   
1  0.304963                      -0.484136                -0.384875   
2  0.304963                      -1.199330                -0.342847   
3 -0.111631                      -1.351886                 0.927490   
4 -0.111631                      -1.017382                 0.626514   

   Number_of_hotel_rooms  Rainfall     Parks  index  Airport_NO  Airport_YES  \
0              -0.352111 -1.302068 -0.473132      0       False         True   
1              -0.166019  0.215194 -0.775225      1        True        False   
2              16.710340 -0.104230 -0.811237      2        True        False   
3              -0.337843  0.454762 -0.680402      3       False         True   
4              -0.333593  1.253320 -1.404723      4        True        False   

   Waterbody_Lake  Waterbody_Lake and River  Waterbody_River  Bus_Terminal_YES  
0           False                     False             True              True  
1            True                     False            False              True  
2           False                     False             True              True  
3            True                     False            False              True  
4            True                     False            False              True

In [36]:
from sklearn.model_selection import train_test_split
import statsmodels
import statsmodels.api as sm
  
X = sm.add_constant(X)
X_train,X_test, y_train, y_test = train_test_split( X,
                                                   target, 
                                                   random_state = 10, 
                                                   test_size = 0.2)
print('X_train',X_train.shape)
print('y_train',y_train.shape)

X_train (398, 21)
y_train (398,)


In [37]:
X.head()

const  index     Price  Resid_Area  Air_Qual  Room_Nums       Age  \
0    1.0      0  0.154792   -1.282444 -0.137094   0.408975 -0.115644   
1    1.0      1 -0.106592   -0.590059 -0.733195   0.189974  0.369741   
2    1.0      2  1.320127   -0.590059 -0.733195   1.276446 -0.260905   
3    1.0      3  1.178544   -1.301354 -0.828226   1.010516 -0.802977   
4    1.0      4  1.483492   -1.301354 -0.828226   1.222407 -0.505369   

   Distance  Teachers  Preportion_of_poor_population  ...  \
0  0.172845  1.462167                      -1.066368  ...   
1  0.476100  0.304963                      -0.484136  ...   
2  0.495053  0.304963                      -1.199330  ...   
3  1.054179 -0.111631                      -1.351886  ...   
4  1.030487 -0.111631                      -1.017382  ...   

   Number_of_hotel_rooms  Rainfall     Parks  index  Airport_NO  Airport_YES  \
0              -0.352111 -1.302068 -0.473132      0       False         True   
1              -0.166019  0.215194 -0.775225      1        True        False   
2              16.710340 -0.104230 -0.811237      2        True        False   
3              -0.337843  0.454762 -0.680402      3       False         True   
4              -0.333593  1.253320 -1.404723      4        True        False   

   Waterbody_Lake  Waterbody_Lake and River  Waterbody_River  Bus_Terminal_YES  
0           False                     False             True              True  
1            True                     False            False              True  
2           False                     False             True              True  
3            True                     False            False              True  
4            True                     False            False              True  

[5 rows x 21 columns]

# Decision tree classifier

In [38]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score


In [39]:
tree_model = DecisionTreeClassifier()
tree_model.fit(X_train, y_train)
predictions = tree_model.predict(X_test)
accuracy = accuracy_score(y_test, predictions)
print("Accuracy:",accuracy)


Accuracy: 0.61


In [40]:
predictions

array([1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0,
       1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0,
       0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1,
       1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0], dtype=int64)

In [41]:
accuracy = accuracy_score(y_test, predictions)
print("Accuracy: ", accuracy)

Accuracy:  0.61


# Random forest Classifier

In [42]:
from sklearn.ensemble import RandomForestClassifier

# Assuming X_train, X_test, y_train, y_test are already defined from train_test_split

# Create a Random Forest classifier
forest_model = RandomForestClassifier(n_estimators=100, random_state=42)

# Fit the model to the training data
forest_model.fit(X_train, y_train)

# Predict on the testing data
predictions = forest_model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, predictions)
print("Accuracy:", accuracy)

Accuracy: 0.61


# Logistic Regression

In [43]:
from sklearn.linear_model import LogisticRegression

# Assuming X_train, X_test, y_train, y_test are already defined from train_test_split

# Create a Logistic Regression classifier
logistic_model = LogisticRegression()

# Fit the model to the training data
logistic_model.fit(X_train, y_train)

# Predict on the testing data
predictions = logistic_model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, predictions)
print("Accuracy:", accuracy)

Accuracy: 0.65


C:\ProgramData\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


# GaussianNB

In [44]:
from sklearn.naive_bayes import GaussianNB

# Assuming X_train, X_test, y_train, y_test are already defined from train_test_split

# Create a Naive Bayes classifier (Gaussian Naive Bayes for continuous features)
naive_bayes_model = GaussianNB()

# Fit the model to the training data
naive_bayes_model.fit(X_train, y_train)

# Predict on the testing data
predictions = naive_bayes_model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, predictions)
print("Accuracy:", accuracy)

Accuracy: 0.59


# KNeighborsClassifier

In [45]:
from sklearn.neighbors import KNeighborsClassifier

# Assuming X_train, X_test, y_train, y_test are already defined from train_test_split
# Also assuming you have imported your dataset and labels as X and y respectively

# Check if data is numeric
if not X_train.applymap(np.isreal).all().all():
    print("Error: Data is not numeric.")

# Check data types
if not X_train.select_dtypes(include=[np.number]).equals(X_train):
    print("Error: Data types are not compatible with scikit-learn.")

# Check data shape
if len(X_train.shape) != 2 or len(X_test.shape) != 2:
    print("Error: Data shape is not correct. Expected 2D arrays.")

# Check for NaN values
if X_train.isnull().values.any() or X_test.isnull().values.any():
    print("Error: Data contains NaN values.")

# Convert DataFrames to NumPy arrays
X_train_np = X_train.values
X_test_np = X_test.values

# Check data contiguity
if not X_train_np.flags.c_contiguous or not X_test_np.flags.c_contiguous:
    print("Error: Data is not contiguous.")

# Create a KNN classifier
k = 5  # number of neighbors
knn_model = KNeighborsClassifier(n_neighbors=k)

# Fit the model to the training data
knn_model.fit(X_train_np, y_train)

# Predict on the testing data
predictions = knn_model.predict(X_test_np)

# Calculate accuracy
accuracy = accuracy_score(y_test, predictions)
print("Accuracy:", accuracy)

Error: Data types are not compatible with scikit-learn.
Error: Data is not contiguous.


C:\Users\HP\AppData\Local\Temp\ipykernel_11580\390295277.py:7: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  if not X_train.applymap(np.isreal).all().all():


Accuracy: 0.56


# Artificial Neural Network

In [46]:
import numpy as np
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# Assuming X_train, X_test, y_train, y_test are already defined from train_test_split

# Convert data types to float32 if necessary
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')

# Normalize the input data
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# Create a Sequential model
model = Sequential()

# Add the input layer and the first hidden layer
model.add(Dense(units=64, activation='relu', input_shape=(X_train.shape[1],)))

# Add additional hidden layers
model.add(Dense(units=64, activation='relu'))

# Add the output layer
model.add(Dense(units=1, activation='sigmoid'))

# Compile the model
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Train the model
model.fit(X_train, y_train, epochs=10, batch_size=32, validation_split=0.2)

# Evaluate the model on the testing data
predictions = model.predict(X_test)
predictions = (predictions > 0.5)  # Convert probabilities to binary predictions
accuracy = accuracy_score(y_test, predictions)
print("Accuracy:", accuracy)

Epoch 1/10


C:\Users\HP\AppData\Roaming\Python\Python312\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


10/10 ━━━━━━━━━━━━━━━━━━━━ 2s 32ms/step - accuracy: 0.4865 - loss: 0.7023 - val_accuracy: 0.6000 - val_loss: 0.6706
Epoch 2/10
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.5564 - loss: 0.6707 - val_accuracy: 0.6000 - val_loss: 0.6582
Epoch 3/10
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.5814 - loss: 0.6539 - val_accuracy: 0.6000 - val_loss: 0.6582
Epoch 4/10
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.6696 - loss: 0.6264 - val_accuracy: 0.6000 - val_loss: 0.6553
Epoch 5/10
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.6434 - loss: 0.6217 - val_accuracy: 0.6250 - val_loss: 0.6514
Epoch 6/10
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.6899 - loss: 0.6080 - val_accuracy: 0.6250 - val_loss: 0.6481
Epoch 7/10
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.6828 - loss: 0.5857 - val_accuracy: 0.6125 - val_loss: 0.6416
Epoch 8/10
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.6990 - loss: 0.5797 - val_accuracy: 0.6125 - val_loss: 0.6416
Ep

# Linear Discriminant Analysis

In [47]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

# Assuming X_train, X_test, y_train, y_test are already defined from train_test_split

# Create an LDA classifier
lda_model = LinearDiscriminantAnalysis()

# Fit the model to the training data
lda_model.fit(X_train, y_train)

# Predict on the testing data
predictions = lda_model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, predictions)
print("Accuracy:", accuracy)

Accuracy: 0.66


# Quadratic Discriminant Analysis

In [48]:
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis

# Assuming X_train, X_test, y_train, y_test are already defined from train_test_split

# Create a QDA classifier
qda_model = QuadraticDiscriminantAnalysis()

# Fit the model to the training data
qda_model.fit(X_train, y_train)

# Predict on the testing data
predictions = qda_model.predict(X_test)

# Calculate accuracy
accuracy = accuracy_score(y_test, predictions)
print("Accuracy:", accuracy)


Accuracy: 0.49


C:\ProgramData\anaconda3\Lib\site-packages\sklearn\discriminant_analysis.py:935: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")
C:\ProgramData\anaconda3\Lib\site-packages\sklearn\discriminant_analysis.py:960: RuntimeWarning: divide by zero encountered in power
  X2 = np.dot(Xm, R * (S ** (-0.5)))
C:\ProgramData\anaconda3\Lib\site-packages\sklearn\discriminant_analysis.py:960: RuntimeWarning: invalid value encountered in multiply
  X2 = np.dot(Xm, R * (S ** (-0.5)))
C:\ProgramData\anaconda3\Lib\site-packages\sklearn\discriminant_analysis.py:963: RuntimeWarning: divide by zero encountered in log
  u = np.asarray([np.sum(np.log(s)) for s in self.scalings_])
